src/util/token.rs

```diff
@@ -0,0 +1,133 @@
+use rand::{distributions::Uniform, rngs::OsRng, Rng};
+use sha2::{Digest, Sha256};
+
+const TOKEN_LENGTH: usize = 32;
+
+pub(crate) struct SecureToken {
+    sha256: Vec<u8>,
+}
+
+impl SecureToken {
+    pub(crate) fn generate(kind: SecureTokenKind) -> NewSecureToken {
+        let plaintext = format!(
+            "{}{}",
+            kind.prefix(),
+            generate_secure_alphanumeric_string(TOKEN_LENGTH)
+        );
+        let sha256 = Sha256::digest(plaintext.as_bytes()).as_slice().to_vec();
+
+        NewSecureToken {
+            plaintext,
+            inner: Self { sha256 },
+        }
+    }
+
+    pub(crate) fn parse(kind: SecureTokenKind, plaintext: &str) -> Option<Self> {
+        // This will both reject tokens without a prefix and tokens of the wrong kind.
+        if SecureTokenKind::from_token(plaintext) != Some(kind) {
+            return None;
+        }
+
+        let sha256 = Sha256::digest(plaintext.as_bytes()).as_slice().to_vec();
+        Some(Self { sha256 })
+    }
+
+    pub(crate) fn sha256(&self) -> &[u8] {
+        &self.sha256
+    }
+}
+
+pub(crate) struct NewSecureToken {
+    plaintext: String,
+    inner: SecureToken,
+}
+
+impl NewSecureToken {
+    pub(crate) fn plaintext(&self) -> &str {
+        &self.plaintext
+    }
+}
+
+impl std::ops::Deref for NewSecureToken {
+    type Target = SecureToken;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+fn generate_secure_alphanumeric_string(len: usize) -> String {
+    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+
+    OsRng
+        .sample_iter(Uniform::from(0..CHARS.len()))
+        .map(|idx| CHARS[idx] as char)
+        .take(len)
+        .collect()
+}
+
+macro_rules! secure_token_kind {
```

Comment:
I find this a bit hard to read. Did you consider using something like `strum` instead of this custom macro? If we'll still go with the custom macro I think it'd be good to add a few more doc comments on it at least :)

Reply:
I don't think it's worth bringing another dependency in just for this. I'll try to add some comments to the macrol.

Reply:
I guess this is also fine to retrofit. I'm approving the PR in the meantime :)